File: backend/app/validation_registry.py
import json
from pathlib import Path

_SHARED_DIR = Path(__file__).resolve().parent.parent.parent / "shared"

with open(_SHARED_DIR / "validations.json") as f:
    _defs = json.load(f)

with open(_SHARED_DIR / "commands.json") as f:
    _commands = json.load(f)

with open(_SHARED_DIR / "providers.json") as f:
    PROVIDERS: dict[str, dict] = json.load(f)

VALID_PROVIDERS: list[str] = ["custom", *PROVIDERS.keys()]


def _resolve_source(source: str) -> list[str]:
    """Resolve 'commands.matchTypes' style references to actual values."""
    file_key, field_key = source.split(".")
    if file_key == "commands":
        return _commands[field_key]
    raise ValueError(f"Unknown source: {source}")


ENUMS: dict[str, list[str]] = _defs["enums"]
SCENARIO_FIELDS: dict[str, dict] = _defs["scenario"]
VALIDATION_FIELDS: dict[str, dict] = _defs["validation"]
URL_CHECK_FIELDS: dict[str, dict] = _defs["urlCheck"]
PARAM_CHECK_FIELDS: dict[str, dict] = _defs["paramCheck"]


def pydantic_field_kwargs(field_def: dict) -> dict:
    """Convert a shared field definition to Pydantic Field() kwargs."""
    kwargs: dict = {}
    if "minLength" in field_def:
        kwargs["min_length"] = field_def["minLength"]
    if "maxLength" in field_def:
        kwargs["max_length"] = field_def["maxLength"]
    if "min" in field_def:
        kwargs["ge"] = field_def["min"]
    if "max" in field_def:
        kwargs["le"] = field_def["max"]
    if field_def.get("positive"):
        kwargs["gt"] = 0
    if "default" in field_def:
        kwargs["default"] = field_def["default"]
    if "minItems" in field_def:
        kwargs["min_length"] = field_def["minItems"]
    return kwargs

File: extension/src/lib/validation-registry.ts
// Validation registry — data sourced from shared/validations.json (single source of truth).

import { z } from "zod/v4";
import validationDefs from "../../../shared/validations.json";
import commandDefs from "../../../shared/commands.json";
import providerDefs from "../../../shared/providers.json";

// --- Types ---

interface FieldDef {
  type: string;
  required?: boolean;
  default?: unknown;
  minLength?: number;
  maxLength?: number;
  min?: number;
  max?: number;
  positive?: boolean;
  minItems?: number;
  source?: string;
  conditionalRequired?: {
    unless: { field: string; equals: string };
    minLength: number;
  };
}

// --- Source resolver ---

function resolveSource(source: string): string[] {
  const [fileKey, fieldKey] = source.split(".");
  if (fileKey === "commands") {
    return (commandDefs as Record<string, unknown>)[fieldKey] as string[];
  }
  throw new Error(`Unknown source: ${source}`);
}

// --- Provider registry ---

export interface ParamSuggestion {
  label: string;
  value: string;
  valueSuggestions: string[];
}

interface ProviderDef {
  name: string;
  urlPatterns: string[];
  paramSuggestions?: ParamSuggestion[];
}

export const PROVIDERS: Record<string, ProviderDef> = providerDefs;
export const PROVIDER_VALUES = ["custom", ...Object.keys(PROVIDERS)] as [string, ...string[]];

// --- Exported constants ---

export const ENUMS = validationDefs.enums;
export const SCENARIO_FIELDS = validationDefs.scenario as Record<string, FieldDef>;
export const URL_CHECK_FIELDS = validationDefs.urlCheck as Record<string, FieldDef>;
export const PARAM_CHECK_FIELDS = validationDefs.paramCheck as Record<string, FieldDef>;

// --- Zod schema builders ---

export function buildStringSchema(def: FieldDef) {
  let s = z.string();
  if (def.minLength !== undefined) s = s.min(def.minLength);
  if (def.maxLength !== undefined) s = s.max(def.maxLength);
  return s;
}

export function buildIntSchema(def: FieldDef) {
  let s = z.number().int();
  if (def.positive) s = s.positive();
  if (def.min !== undefined) s = s.min(def.min);
  if (def.max !== undefined) s = s.max(def.max);
  if (def.default !== undefined) s = s.default(def.default as number);
  return s;
}

export function buildEnumSchema(def: FieldDef) {
  if (!def.source) throw new Error("Enum field requires a 'source'");
  const values = resolveSource(def.source) as [string, ...string[]];
  return z.enum(values);
}

